jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/NodeInstanceImpl.java

```diff
@@ -257,113 +264,119 @@ protected void triggerCompleted(String type, boolean remove) {
-                }   
+                }
                 return;
-        	} else {
-        		connections = node.getOutgoingConnections(type); 
-        	}
+            } else {
+                connections = node.getOutgoingConnections(type);
+            }
         }
         if (connections == null || connections.isEmpty() ) {
-        	boolean hidden = false;
-        	Node currentNode = getNode();
-        	if (currentNode != null && currentNode.getMetaData().get("hidden") != null) {
-        		hidden = true;
-        	}
-        	InternalKnowledgeRuntime kruntime = getProcessInstance().getKnowledgeRuntime();
-        	if (!hidden) {
-        		((InternalProcessRuntime) kruntime.getProcessRuntime())
-        			.getProcessEventSupport().fireBeforeNodeLeft(this, kruntime);
-        	}
-        	// notify container
+            boolean hidden = isHidden();
+            InternalKnowledgeRuntime kruntime = getProcessInstance().getKnowledgeRuntime();
+            if (!hidden) {
+                ((InternalProcessRuntime) kruntime.getProcessRuntime())
+.getProcessEventSupport().fireBeforeNodeLeft(this, kruntime);
+            }
+            // notify container
             ((org.jbpm.workflow.instance.NodeInstanceContainer) getNodeInstanceContainer())
-        		.nodeInstanceCompleted(this, type);
+                .nodeInstanceCompleted(this, type);
             if (!hidden) {
-            	((InternalProcessRuntime) kruntime.getProcessRuntime())
-            		.getProcessEventSupport().fireAfterNodeLeft(this, kruntime);
+                ((InternalProcessRuntime) kruntime.getProcessRuntime())
+.getProcessEventSupport().fireAfterNodeLeft(this, kruntime);
             }
         } else {
-        	Map<org.jbpm.workflow.instance.NodeInstance, String> nodeInstances = 
-        		new HashMap<org.jbpm.workflow.instance.NodeInstance, String>();
-        	for (Connection connection: connections) {
-        		nodeInstances.put(followConnection(connection), connection.getToType());
-        	}
-        	for (Map.Entry<org.jbpm.workflow.instance.NodeInstance, String> nodeInstance: nodeInstances.entrySet()) {
-	        	// stop if this process instance has been aborted / completed
-	        	if (((org.jbpm.workflow.instance.NodeInstanceContainer) getNodeInstanceContainer()).getState() != ProcessInstance.STATE_ACTIVE) {
-	        		return;
-	        	}
-	    		triggerNodeInstance(nodeInstance.getKey(), nodeInstance.getValue());
-	        }
-        }
-    }
-    
+            Deque<TriggerAction> nodeInstances = new ArrayDeque<TriggerAction>(connections.size());
+            for (Connection connection: connections) {
+                nodeInstances.add(new TriggerAction(followConnection(connection), connection.getToType()));
+            }
+            Iterator<TriggerAction> iter;
+            if( isStackless() ) {
+                iter = nodeInstances.descendingIterator();
```

Comment:
similar here, could this be hidden by like a method in NodeInstanceImpl that simply returns Iterator and the change what kind is hidden in that method? That way it will be more readable and less cleanup when moving to single engine mode only.